### PR TITLE
Implement DownsampleFactorMax.R_op 

### PR DESCRIPTION
**Pull request by rodrigob**
This is a PR to attack issue #305.
Imported what seemed relevant from pascanur:downsampling_rop branch (PR #622).
When comparing the two code bases, the pascanur branch seems quite archaic.

I am not familiar with the Theano code base, so I did my best pythonic guesses. I most certainly introduced bugs. This code will need heavy testing and/or upgrades.

`test_rop.py` includes a test for the new feature, and it currently fails.

```
....SES..................
======================================================================
ERROR: test_downsample (theano.tests.test_rop.test_RopLop)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/benenson/code/references/Theano/theano/tests/test_rop.py", line 274, in test_downsample
    yv = tensor.Rop(out, input, eval_p)
  File "/home/benenson/code/references/Theano/theano/gradient.py", line 293, in Rop
    _traverse(out.owner)
  File "/home/benenson/code/references/Theano/theano/gradient.py", line 256, in _traverse
    _traverse(inp.owner)
  File "/home/benenson/code/references/Theano/theano/gradient.py", line 288, in _traverse
    seen_nodes[node] = op.R_op(node.inputs, same_type_eval_points)
  File "/home/benenson/code/references/Theano/theano/tensor/signal/downsample.py", line 346, in R_op
    st=self.st, padding=self.padding)
TypeError: __init__() got an unexpected keyword argument 'padding'

----------------------------------------------------------------------
```

I had added `st=self.st, padding=self.padding` because this are new arguments present on the other downsampling operators. Not sure how/if should be present in Rop.

Anyone interested on helping for this PR is most welcome. Theano internals look quite gloomy.

### CODE REVIEW

**Comment by rodrigob** [comment]
@pascanur  any interest on revisiting this code ?

**Comment by rodrigob** [comment]
Added a bit more code. Now the unit test fails with a C compilation error.
This time I pass the torch to @nouiz  . Good luck with figuring out what is wrong here...

**Comment by nouiz** [comment]
Thanks for advancing this. I don't have the time this week. Maybe next week.

If the problem is just in the c code, you can disable it by raising
NotImplementedError in the function
Le 21 sept. 2015 16:18, "Rodrigo Benenson" <notifications@github.com> a
écrit :

**Comment by rodrigob** [comment]
Adding NotImplementedError to the cpu c_code (plus a couple of minor fixed) make the unit test pass. Indeed.

From what I can see this is the todo list on this PR.

- [ ] Make sure the unit test is testing enough
- [ ] `DownsampleFactorMaxRop.st` and `.padding` are currently ignored. Is this fine ?
- [ ] Make CPU c_code work as intended
- [ ] Make GPU c_code work as intended

@nouiz looking forward for your help on getting these fixed.
(since I am working "in the blind" regarding how Theano internals work)

File: theano/sandbox/cuda/opt.py [hunk 1]
```diff
@@ -1867,6 +1867,27 @@ def local_gpu_downsample_factor_max_grad(node):
...
+@local_optimizer([downsample.DownsampleFactorMaxRop])
+def local_gpu_downsample_factor_max_rop(node):
+    if (isinstance(node.op, downsample.DownsampleFactorMaxRop) and
+        node.op.ds == node.op.st):
+
+        assert node.op.__props__ == ('ds', 'ignore_border', 'st', 'padding',
+                                     'mode')
+        if (node.op.padding != (0, 0) or
+            node.op.mode != 'max' or
+            node.op.st != node.op.ds):
```
Comment (nouiz) [inline theano/sandbox/cuda/opt.py]: that seem the good update for now compared to the original code in the other PR.

File: theano/tensor/signal/downsample.py [hunk 2]
```diff
@@ -333,6 +333,19 @@ def grad(self, inp, grads):
...
                                         x, gz)]
+
+    def R_op(self, inputs, eval_points):
+        # R_op can receive None as eval_points.
+        # That mean there is no diferientiable path through that input
+        # If this imply that you cannot compute some outputs,
+        # return None for those.
+        if eval_points[0] is None:
+            return [None]
+        rop = DownsampleFactorMaxRop(self.ds,
```
Comment (nouiz) [inline theano/tensor/signal/downsample.py]: Here, until padding and strides are supported raise NotImplementedError when padding isn't 0 and strides isn't the same as ds.

Do you need those feature?

File: theano/tensor/signal/downsample.py [hunk 3]
```diff
@@ -1146,3 +1159,206 @@ def local_average_pool_grad(node):
...
+            return TypeError('Expected tensor4')
+        # TODO: consider restrucing the dtype?
+        x = tensor.as_tensor_variable(x)
+        eval_point = tensor.as_tensor_variable(eval_point)
+        return gof.Apply(self, [x, eval_point], [x.type()])
+
+    def perform(self, node, inp, out):
+        x, ex = inp
+        z, = out
+        if len(x.shape) != 4:
```
Comment (nouiz) [inline theano/tensor/signal/downsample.py]: You need to update this method to do the right implementation when padding != 0 and when strides != ds. If not, raise an error in the __init__ for the not implemented case.

This is python code, the not efficient version. If it help you, you can explicitly pad the inputs if that help you make the support for this case.

File: theano/tests/test_rop.py [hunk 4]
```diff
@@ -259,6 +259,33 @@ def test_rebroadcast(self):
...
 
+    def test_downsample(self):
+        rng = numpy.random.RandomState(utt.fetch_seed())
+        maxpoolshps = ((1,1),(3,2),(2,3))
+        vx = (rng.rand(2,3,3,4) * 2.0).astype(theano.config.floatX)
+        vv = (rng.rand(2,3,3,4) * 2.0).astype(theano.config.floatX)
+        input = theano.shared(vx)
+        eval_p = theano.shared(vv)
+        for maxpoolshp in maxpoolshps:
+            for ignore_border in [False, True]:
```
Comment (nouiz) [inline theano/tests/test_rop.py]: If you add the padding/strides support, then you need to add extra loop to tests those condition.

Also, I think that now we have a better automatic test that is documented here:

http://www.deeplearning.net/software/theano/tutorial/extending_theano.html#testing-the-rop

**Comment by nouiz** [comment]
I have a PR to your PR with updates;

https://github.com/rodrigob/Theano/pull/1

**Comment by nboley** [comment]
Have you made any progress on this? I'm currently blocked on this - and would be happy to help if you can tell me where you're at.

Thanks!

**Comment by nouiz** [comment]
You can use this PR. It just that it would need the tests to be changed to
use the Rop tests infrastructure and I don't have the time to do it.

If you or someone else would like to do that, it would be great.

**Comment by nboley** [comment]
I'll give it a shot - thanks for the quick response. 

Also, unless I'm missing something, Rop still needs to be implemented for MaxPoolGrad.

**Comment by nouiz** [comment]
Continued in https://github.com/Theano/Theano/pull/4781

**Pull request closed without merging**